lib/evaluate/type.h

```diff
@@ -0,0 +1,107 @@
+// total bit width and real precision.
+
+#include "complex.h"
+#include "integer.h"
+#include "logical.h"
+#include "real.h"
+
+namespace Fortran::evaluate::type {
+
+enum class Classification { Integer, Real, Complex, Character, Logical };
+
+template<int KIND> struct Integer {
+  static constexpr Classification classification{Classification::Integer};
+  static constexpr int kind{KIND};
+  static constexpr bool hasLen{false};
+  using ValueType = value::Integer<8 * kind>;
+};
+
+template<int KIND> struct Real;
+template<> struct Real<2> {
+  static constexpr Classification classification{Classification::Real};
+  static constexpr int kind{2};
+  static constexpr bool hasLen{false};
+  using ValueType = value::Real<typename Integer<kind>::ValueType, 11>;
+};
+template<> struct Real<4> {
+  static constexpr Classification classification{Classification::Real};
+  static constexpr int kind{4};
+  static constexpr bool hasLen{false};
+  using ValueType = value::Real<typename Integer<kind>::ValueType, 24>;
+};
+template<> struct Real<8> {
+  static constexpr Classification classification{Classification::Real};
+  static constexpr int kind{8};
+  static constexpr bool hasLen{false};
+  using ValueType = value::Real<typename Integer<kind>::ValueType, 53>;
+};
+template<> struct Real<10> {
+  static constexpr Classification classification{Classification::Real};
+  static constexpr int kind{10};
+  static constexpr bool hasLen{false};
+  using ValueType = value::Real<value::Integer<80>, 64, false>;
+};
+template<> struct Real<16> {
+  static constexpr Classification classification{Classification::Real};
+  static constexpr int kind{16};
+  static constexpr bool hasLen{false};
+  using ValueType = value::Real<typename Integer<kind>::ValueType, 112>;
+};
+
+// The KIND type parameter on COMPLEX is the kind of each of its components.
+template<int KIND> struct Complex {
+  static constexpr Classification classification{Classification::Complex};
+  static constexpr int kind{KIND};
+  static constexpr bool hasLen{false};
+  using ValueType = value::Complex<typename Real<kind>::ValueType>;
+};
+
+template<int KIND> struct Logical {
+  static constexpr Classification classification{Classification::Logical};
+  static constexpr int kind{KIND};
+  static constexpr bool hasLen{false};
+  using ValueType = value::Logical<8 * kind>;
+};
+
+template<int KIND> struct Character {
+  static constexpr Classification classification{Classification::Character};
+  static constexpr int kind{KIND};
+  static constexpr bool hasLen{true};
+  using ValueType = std::uint8_t[kind];  // TODO: ?
+};
+
+// Default REAL just simply has to be IEEE-754 single precision today.
+// It occupies one numeric storage unit by definition.  The default INTEGER
+// and default LOGICAL intrinsic types also have to occupy one numeric
+// storage unit, so their kinds are also forced.  Default COMPLEX occupies
+// two numeric storage units.
+
+using DefaultReal = Real<4>;
```

Comment:
How will this work with the command line option that changes default real to be Real<8>?

Reply:
We need to discuss that option and whether it's valid in the context of the current standard and what its consequences would be for storage association -- would it also change the byte size of a single numeric storage unit, and hence the size of default integer and logical?

Reply:
PGI, gfortran, and ifort have -r8 and similar options that control other default real and integer (and logical) sizes.

Reply:
Yes, I know.  What I don't know is the extent to which those options pose problems for numeric storage association, which requires that default real and default integer and default logical all occupy a single numeric storage unit by definition.  I also don't know whether they affect explicitly kinded declarations (e.g. `REAL(KIND=4)` gets silently modified) or just default-kinded declarations and constants.  I won't be able to answer your original question without doing more research.